src/ct_eval.cpp

```diff
@@ -211,6 +232,32 @@ MonoId mono_type_from_ast(TypedAST::TypedAST* ast, TypeChecker& tc){
 	}
 }
 
+TypedAST::Constructor* constructor_from_ast(
+    TypedAST::TypedAST* ast, TypeChecker& tc, TypedAST::Allocator& alloc) {
+	MetaTypeId meta = tc.m_core.m_meta_core.find(ast->m_meta_type);
+	auto constructor = alloc.make<TypedAST::Constructor>();
+	constructor->m_syntax = ast;
+
+	if (meta == tc.meta_monotype()) {
+		auto mono_handle = static_cast<TypedAST::MonoTypeHandle*>(
+		    ct_eval(ast, tc, alloc));
+		constructor->m_mono = mono_handle;
+	} else if (meta == tc.meta_constructor()) {
+		assert(ast->type() == TypedASTTag::AccessExpression);
+
+		auto access = static_cast<TypedAST::AccessExpression*>(ast);
+		auto mono_handle = static_cast<TypedAST::MonoTypeHandle*>(
+		    ct_eval(access->m_object, tc, alloc));
+
+		constructor->m_mono = mono_handle;
+		constructor->m_id = access->m_member;
```

Comment:
What would happen if we haven't yet evaluated the type func when we reach this point?

example:

```
x := my_union(<>).a { 10 };
my_union := union {
  a : int(<>);
};
```

Reply:
If this is wrong as I think it is, the fix would be to use unification, kinda like the thing we do when we access a member of a record, but we potentially don't know the record yet.

Reply:
I did not think about this case. I will add a test and make it work

Reply:
Doing the example you suggested seems to work. I think it works because `match_identifiers` takes care of it, so by the time we compile, the constructor creator calls the mono type creator, which calls the type function creator that does the rest.

Reply:
Oh I'm surprised. I'll give your patch a more thorough reading in a bit then

Reply:
Ah yes, recursive typefuncs will probably expose a bunch of hidden assumptions throughout the codebase

Reply:
I don't understand why it works yet, but it looks like it does... huh